python_modules/dagster/dagster/_core/definitions/external_asset_graph.py

```diff
@@ -122,106 +93,216 @@ def from_repository_handles_and_external_asset_nodes(
-        auto_materialize_policies_by_key = {}
-        backfill_policies_by_key = {}
-        keys_by_atomic_execution_unit_id: Dict[str, Set[AssetKeyOrCheckKey]] = defaultdict(set)
         repo_handles_by_key = {
             node.asset_key: repo_handle
             for repo_handle, node in repo_handle_external_asset_nodes
             if not node.is_source or node.is_observable
         }
-        job_names_by_key = {
-            node.asset_key: node.job_names
-            for _, node in repo_handle_external_asset_nodes
-            if not node.is_source or node.is_observable
+
+        # Split the nodes into materializable, observable, and unexecutable nodes. Observable and
+        # unexecutable nodes represent both source and external assets.
+        materializable_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        observable_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        unexecutable_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        for repo_handle, node in repo_handle_external_asset_nodes:
+            if node.is_source and node.is_observable:
+                observable_node_pairs.append((repo_handle, node))
+            elif node.is_source:
+                unexecutable_node_pairs.append((repo_handle, node))
+            else:
+                materializable_node_pairs.append((repo_handle, node))
+
+        asset_nodes_by_key = {}
+
+        # It is possible for multiple nodes to exist that share the same key. This is invalid if
+        # more than one node is materializable or if more than one node is observable. It is valid
+        # if there is at most one materializable node and at most one observable node, with all
+        # other nodes unexecutable. The asset graph will receive only a single `ExternalAssetNode`
+        # representing the asset. This will always be the materializable node if one exists; then
+        # the observable node if it exists; then finally the first-encountered unexecutable node.
+        for repo_handle, node in materializable_node_pairs:
+            if node.asset_key in asset_nodes_by_key:
+                check.failed("Found two materialization nodes with the same asset key")
+            asset_nodes_by_key[node.asset_key] = node
+
+        for repo_handle, node in observable_node_pairs:
+            if node.asset_key in asset_nodes_by_key:
+                current_node = asset_nodes_by_key[node.asset_key]
+                if current_node.is_observable:
+                    check.failed("Found two observable source nodes with the same asset key")
+                asset_nodes_by_key[node.asset_key] = current_node._replace(is_observable=True)
+            else:
+                asset_nodes_by_key[node.asset_key] = node
+
+        for repo_handle, node in unexecutable_node_pairs:
+            if node.asset_key in asset_nodes_by_key:
+                continue
+            asset_nodes_by_key[node.asset_key] = node
```

Comment:
extracting this function into its own block might be nice

Reply:
Disagree here as it's tightly coupled to the preceding block that splits nodes into materializable/observable/unexecutable, and these two blocks together comprise basically the entire function.